lib/location/method_gnss.c

```diff
@@ -0,0 +1,835 @@
+#include <modem/lte_lc.h>
+#include <modem/at_cmd.h>
+#include <nrf_modem_gnss.h>
+#include <nrf_errno.h>
+#include "loc_core.h"
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+#include <net/nrf_cloud_rest.h>
+#include <net/nrf_cloud_agps.h>
+#include <stdlib.h>
+#endif
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+#include <net/nrf_cloud_rest.h>
+#include <net/nrf_cloud_agps.h>
+#include <net/nrf_cloud_pgps.h>
+#include <pm_config.h>
+#endif
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+#include <date_time.h>
+#endif
+
+LOG_MODULE_DECLARE(location, CONFIG_LOCATION_LOG_LEVEL);
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+BUILD_ASSERT(IS_ENABLED(CONFIG_DATE_TIME));
+#endif
+
+/* Maximun waiting time before GNSS is started regardless of RRC or PSM state. This prevents
+ * Location library from getting stuck indefinitely if the application keeps LTE connection
+ * constantly active.
+ */
+#define SLEEP_WAIT_BACKSTOP 5
+#if !defined(CONFIG_NRF_CLOUD_AGPS)
+/* range 10240-3456000000 ms, see AT command %XMODEMSLEEP */
+#define MIN_SLEEP_DURATION_FOR_STARTING_GNSS 10240
+#define AT_MDM_SLEEP_NOTIF_START "AT%%XMODEMSLEEP=1,%d,%d"
+#define AT_MDM_SLEEP_NOTIF_STOP "AT%XMODEMSLEEP=0" /* not used at the moment */
+#endif
+#if (defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS))
+#define AGPS_REQUEST_RECV_BUF_SIZE 3500
+#define AGPS_REQUEST_HTTPS_RESP_HEADER_SIZE 400
+#endif
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+#define NUM_PREDICTIONS CONFIG_NRF_CLOUD_PGPS_NUM_PREDICTIONS
+#define PREDICTION_PERIOD CONFIG_NRF_CLOUD_PGPS_PREDICTION_PERIOD
+#endif
+
+extern location_event_handler_t event_handler;
+extern struct loc_event_data current_event_data;
+
+struct k_work_args {
+	struct k_work work_item;
+	struct loc_gnss_config gnss_config;
+};
+
+static struct k_work_args method_gnss_start_work;
+static struct k_work method_gnss_pvt_work;
+
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+static struct k_work method_gnss_agps_request_work;
+#if !defined(CONFIG_NRF_CLOUD_MQTT)
+static char agps_data_buf[AGPS_REQUEST_RECV_BUF_SIZE];
+#endif
+#endif
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+static struct k_work method_gnss_pgps_request_work;
+static struct k_work method_gnss_manage_pgps_work;
+static struct k_work method_gnss_notify_pgps_work;
+static struct nrf_cloud_pgps_prediction *prediction;
+static struct gps_pgps_request pgps_request;
+#endif
+
+static int fixes_remaining;
+static bool running;
+static K_SEM_DEFINE(entered_psm_mode, 0, 1);
+static K_SEM_DEFINE(entered_rrc_idle, 0, 1);
+
+#if (defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS))
+static struct nrf_modem_gnss_agps_data_frame agps_request;
+#if !defined(CONFIG_NRF_CLOUD_MQTT)
+static char rest_api_recv_buf[CONFIG_NRF_CLOUD_REST_FRAGMENT_SIZE +
+			      AGPS_REQUEST_HTTPS_RESP_HEADER_SIZE];
+static char jwt_buf[600];
+#endif
+#endif
+
+#if defined(CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL)
+static struct k_work method_gnss_agps_cb_work;
+static void method_gnss_agps_cb_work_fn(struct k_work *item);
```

Comment:
I was initially thinking that I'd like the name of the work and function to include word external in some form (ext or external). I'm seeing that cb probably tries to illustrate the same thing but would ext be clearer as it refers to the config flag?